tests/run-with-compiler-custom-args/staged-streams_1.scala

```diff
@@ -0,0 +1,697 @@
+    * and a stepper function:
+    *
+    * {{
+    *   trait Stream[+A]
+    *   case class Unfold[S, +A](state: S, step: (S) => Option[(S, A)]) extends Stream[+A]
+    * }}
+    *
+    * The latter transforms the state and returns either the end-of-the-stream or a value and
+    * the new state. The existential quantification over the state keeps it private: the only permissible operation is
+    * to pass it to the function.
+    *
+    * Note: in `Producer` the elements are not pulled but the step accepts a continuation.
+    *
+    * A Producer defines the three basic elements of a loop structure:
+    * - `init` contributes the code before iteration starts
+    * - `step` contributes the code during execution
+    * - `hasNext` contributes the code of the boolean test to end the iteration
+    *
+    * @tparam A type of the collection element. Since a `Producer` is polymorphic it handles `Expr` values, we
+    *           can pack together fragments of code to accompany each element production (e.g., a variable incremented
+    *           during each transformation)
+    */
+  trait Producer[A] { self =>
+    type St
+    val card: Cardinality
+
+    /** Initialization method that defines new state, if needed by the combinator that this producer defines.
+      *
+      * @param  k the continuation that is invoked after the new state is defined in the body of `init`
+      * @return expr value of unit per the CPS-encoding
+      */
+    def init(k: St => Expr[Unit]): Expr[Unit]
+
+    /** Step method that defines the transformation of data.
+      *
+      * @param  st the state needed for this iteration step
+      * @param  k  the continuation that accepts each element and proceeds with the step-wise processing
+      * @return expr value of unit per the CPS-encoding
+      */
+    def step(st: St, k: (A => Expr[Unit])): Expr[Unit]
+
+    /** The condition that checks for termination
+      *
+      * @param  st the state needed for this iteration check
+      * @return the expression for a boolean
+      */
+    def hasNext(st: St): Expr[Boolean]
+  }
+
+  trait Cardinality
+  case object AtMost1 extends Cardinality
+  case object Many extends Cardinality
+
+  trait StagedStream[A]
+  case class Linear[A](producer: Producer[A]) extends StagedStream[A]
+  case class Nested[A, B](producer: Producer[B], nestedf: B => StagedStream[A]) extends StagedStream[A]
+
+  case class Stream[A: Type](stream: StagedStream[Expr[A]]) {
+
+    /** Main consumer
+      *
+      * Fold accumulates the results in a variable and delegates its functionality to `foldRaw`
+      *
+      * @param z   the accumulator
+      * @param f   the zipping function
+      * @tparam W  the type of the accumulator
+      * @return
+      */
+     def fold[W: Type](z: Expr[W], f: ((Expr[W], Expr[A]) => Expr[W])): Expr[W] = {
+       Var(z) { s: Var[W] => '{
+           ~foldRaw[Expr[A]]((a: Expr[A]) => '{
+             ~s.update(f(s.get, a))
+           }, stream)
```

Comment:
Is there any difference between `'{ ~s.update(…) }` and `s.update(…)`?

Reply:
It is the same

Reply:
Though I would not change it. This kind of small additions have been useful to find bugs in the implementation. I would keep it like this for as a regression test.

Reply:
However, nice catch @julienrf!